### PR TITLE
#167733644 signup/Registration Validation Errors

### PR DESCRIPTION
**Pull request by dchima**
### What does this PR do?
Validate user inputs and throws errors whenever user does not meet input requirements.

### Description of Task to be completed?
- Install needed validation depenencies (e.g. @hapi/joi package)
- Create user validation middleware
-  Add validation class for user input at signup
- Update routes
- Write the test for validations
- Test validations (on success and on failure)

### How should this be manually tested?
- Clone repository
- update or copy .env file from .env.sample
- Run `npm run prestart` to create tables in your database.
- Run  npm run start:dev on local root directory
- Test router on Postman with POST: 'localhost:3000/api/auth/signup'
- Run a test case JSON script on Postman body, eg:
```
{
	"firstName": "Tony",
	"lastName": "Marshal",
	"email": "tony@marshal.com",
	"password": "tmonarqA1.",
	"gender": "male",
	"street": "Abage",
	"city": "Lekki",
	"state": "Lagos",
	"country": "Nigeria",
	"birthdate": "2009-08-20",
	"phoneNumber": "(+234) 098-3442-1889",
	"companyName": "Jubilee"
}
```
- Omit some fields to check that the validation script works on inputs not given by the user.

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#167733644](https://www.pivotaltracker.com/story/show/167733644)

### CODE REVIEW

**Comment by dchima** [comment]
Forgot to remove the loggers I used for testing. I'll do that in the next commit

File: src/routes/api/users.js [hunk 1]
```diff
@@ -1,8 +1,9 @@
 import express from 'express';
 import UserController from '../../controllers/index';
+import userValidation from '../../validation';
 
 const userRouter = express.Router();
 
-userRouter.post('/auth/signup', UserController.createUser);
+userRouter.post('/auth/signup', userValidation.signup, UserController.createUser);
```
Comment (Segun-Ogundipe) [inline src/routes/api/users.js]: This should be in the middleware folder

Comment (Segun-Ogundipe) [inline src/routes/api/users.js]: Should be in middleware folder

File: src/utils/response.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Response function for error / success returns
```
Comment (Segun-Ogundipe) [inline src/utils/response.js]: A response object was previously added to the project.

File: src/validation/userValidation.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+export default class userValidation {
+  /**
+     * Validates user paramenters upon registration
+     *
+     * @param {object} req - The request from the endpoint.
+     * @param {object} res - The response returned by the method.
+     * @param {object} next - the returned values going into the next operation.
+     * @returns {object} - returns an object (error or response).
+     */
+  static async signup(req, res, next) {
```
Comment (Segun-Ogundipe) [inline src/validation/userValidation.js]: Validations in the validation folder should not be used as a middleware. Validation methods should be created here and imported in a middleware file.

File: src/validation/userValidation.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
+import joi from '@hapi/joi';
+import passwordComplexity from 'joi-password-complexity';
+import { response, ApiResponse } from '../utils';
+import UserService from '../services/index';
+
+// password complexity object
+const complexityOPtions = {
```
Comment (King-Success) [inline src/validation/userValidation.js]: notice typo in complexityOptions, P should be lower case

**Review by Segun-Ogundipe — CHANGES_REQUESTED** [review]
Looks good.
There's no Travis build presently running. Travis build should be fixed before merging

File: src/utils/response.js (hunk 2 above)
Comment (dchima) [reply src/utils/response.js]: I'll remove the redundant response object

**Review by King-Success — APPROVED** [review]

File: src/middlewares/index.js [hunk 5]
```diff
@@ -1,3 +1,4 @@
 import TokenMiddleware from './TokenMiddleware';
+import userMiddleware from './userMiddleware';
 
-export default TokenMiddleware;
+export { TokenMiddleware, userMiddleware };
```
Comment (Segun-Ogundipe) [inline src/middlewares/index.js]: The TTL said we should prefer default export

File: src/middlewares/userMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+export default class ValidationMiddleware {
+/**
+     * Middleware for user validation during signup/registration
+     * @param {object} req - The request from the endpoint.
+     * @param {object} res - The response returned by the method.
+     * @param {object} next - the returned values going into the next operation.
+     * @returns {object} - returns an object (error or response).
+     */
+  static onSignup(req, res, next) {
+    userValidation.signup(req, res, next);
```
Comment (Segun-Ogundipe) [inline src/middlewares/userMiddleware.js]: Uservalidation should either throw an error in the form of ApiError when there's an error in validation or return true when all validation passes.
Sending back responses to the user should be done in the middleware.
The method in the middleware could be renamed to be more descriptive e.g `static validateSignup()`

File: src/validation/userValidation.js [hunk 7]
```diff
@@ -39,7 +39,7 @@ export default class userValidation {
         .label('last name is too short'),
       email: joi.string().email().required()
         .label('invalid email format'),
-      password: new passwordComplexity(complexityOPtions).required()
+      password: new passwordComplexity(complexityOptions).required()
         .label('invalid password format'),
       gender: joi.string().valid('male', 'female').required()
```
Comment (Segun-Ogundipe) [inline src/validation/userValidation.js:46]: The error messages are not descriptive.
For instance, the same error is being sent to the user if the password was not provided or if the format is invalid
The error should be specific. `password is required` when the password is missing `password should include...` when the format is wrong

**Review by Segun-Ogundipe — CHANGES_REQUESTED** [review]
Good job with separating the concerns!!!
I left a few comments on some files.
In the validation, an error should be thrown when validation fails with a descriptive error message. This could be achieved with `new ApiError(400, error.details)` and `return true` when validation passes, then in the middleware, you could:
```try{
       const isValid = userValidation(req.body);

       if(isValid === true) next;
   } catch (error) {
     res.status(error.status || 500).json(new ApiResponse(error.status || 500, error.message));
   }
```
This way, you won't have to send responses in the validator

File: src/middlewares/userMiddleware.js (hunk 6 above)
Comment (dchima) [reply src/middlewares/userMiddleware.js]: The error responses are already handled in the userValidation call. I don't see why this has to be done specifically in the middleware class. All error handling occurs in the validation class already (on success and on failure).

File: src/validation/userValidation.js (hunk 7 above)
Comment (dchima) [reply src/validation/userValidation.js:46]: I'll do just that. However, this doesn't need much recoding, just a change on how descriptive the password error is. If a password isn't inputted, a label like "Password is required. It should be more than 8 characters, and should include at least a capital letter, and a number" should cover both cases in a single error message.

**Comment by dchima** [comment]
> Good job with separating the concerns!!!
> I left a few comments on some files.
> In the validation, an error should be thrown when validation fails with a descriptive error message. This could be achieved with `new ApiError(400, error.details)` and `return true` when validation passes, then in the middleware, you could:
> 
> ```
>        const isValid = userValidation(req.body);
> 
>        if(isValid === true) next;
>    } catch (error) {
>      res.status(error.status || 500).json(new ApiResponse(error.status || 500, error.message));
>    }
> ```
> 
> This way, you won't have to send responses in the validator

Alright, I'll incorporate this in the next commit tonight.

File: src/middlewares/index.js (hunk 5 above)
Comment (dchima) [reply src/middlewares/index.js]: Alright, I'll make the changes.

**Review by Segun-Ogundipe — APPROVED** [review]

**Pull request merged**